roachtest: add multi-region mixed-version test

renatolabs opened the pull request:
This commit adds a `multi-region/mixed-version` test that creates a larger-sized cluster (4 regions, 10 nodes per region) and runs a constant background TPCC worload (with `tolerate-errors`), along with a short TPCC workload that runs in mixed-version state (without `tolerate-errors`).

With this test, we exercise our ability to perform several cluster upgrades in MR clusters. In the future, we plan to extend this test to include other kinds of randomized testing as well.

Fixes: #114803

Release note: None

**roachtest: change TPCC functions to take a logger instance**
This makes it easier to organize logs for a complex test by allowing
the caller to inject the logger instance that should be used in a
particular call to `runTPCC`. More immediately, we make use of this
change in the recently introduced `multi-region/mixed-version` test,
and pass the step's logger to those functions.

Commit:
roachtest: add multi-region mixed-version test

This commit adds a `multi-region/mixed-version` test that creates a
larger-sized cluster (4 regions, 10 nodes per region) and runs a
constant background TPCC worload (with `tolerate-errors`), along with
a short TPCC workload that runs in mixed-version state (without
`tolerate-errors`).

With this test, we exercise our ability to perform several cluster
upgrades in MR clusters. In the future, we plan to extend this test to
include other kinds of randomized testing as well.

Fixes: #114803

Release note: None

Signed-off-by: Renato Costa <renato@cockroachlabs.com>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118293)
<!-- Reviewable:end -->

Commit:
roachtest: change TPCC functions to take a logger instance

This makes it easier to organize logs for a complex test by allowing
the caller to inject the logger instance that should be used in a
particular call to `runTPCC`. More immediately, we make use of this
change in the recently introduced `multi-region/mixed-version` test,
and pass the step's logger to those functions.

Epic: none

Release note: None

renatolabs commented:
I'm trying to convince myself that this test could have caught issues like #113908, but I'm not there yet.

General information about the current test:

* 4 regions, 10 nodes per region
* Very little work done in the cluster: CPU utilization is almost constantly less than 10% across all nodes.
* Basically no data: ~3GB.

While basic, this does seem to mirror the setup in https://github.com/cockroachlabs/support/issues/2685#issuecomment-1788125904 (except for node count -- see below).

****

The upgrade is given 30 minutes to complete. If I comment out `AlwaysUseLatestPredecessors` and run the test with a plan that involves upgrading from 22.2.6 to 23.1.11, the upgrade times out. But by the time it times out, the cluster version is at `22.2-upgrading-to-23.1-step-090`. So it's not done yet, but it's relatively close to finishing.

If I run the test as-is, the upgrade being performed (at the time of writing) will always be 22.2.17 to 23.1.14 (which does include #113996). In this case, the upgrade takes 21 minutes to finish.

****

@nvanbenschoten @fqazi @kvoli: do these numbers look reasonable to you? The original issue seemed to indicate that the upgrade was "stuck", whereas I feel like the 23.1.11 run of this test could have finished if I gave it 10 more minutes. Could this be a matter of cluster size? I could/will try running this on an 80-node cluster but thought I'd get your perspective on this for now.

FWIW, I'm intentionally giving very little work to this cluster because we plan to introduce a `sqlancer` workload in this test very soon, and that will definitely increase utilization.

renatolabs commented:
FWIW, yes, the distinction between 23.1.14 and older versions is much easier to see in with 80 nodes.

renatolabs commented:
@cockroachdb/test-eng ready for review!

srosenberg commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -267,21 +286,32 @@ func runTPCC(ctx context.Context, t test.Test, c cluster.Cluster, opts tpccOptio
+			// test already failed, or that the test finished (and therefore
+			// this function should not return an error).
+			if ctx.Err() != nil && opts.Duration == 0 {
+				return nil
```
I couldn't immediately find examples of tests running the workload indefinitely (i.e., `Duration == 0`).

renatolabs replied:
Yep, the background tpcc workload introduced in this PR should be the only one (hence this logic needed to be added).

srosenberg commented on pkg/cmd/roachtest/tests/mixed_version_multi_region.go:
```diff
@@ -0,0 +1,155 @@
+	const (
+		nodesPerRegion                  = 20
+		backgroundWarehousesPerRegion   = 30
+		mixedVersionWarehousesPerRegion = 50
```
These seem fairly low for a region that has 20 nodes. I'm curious how these were chosen?

renatolabs replied:
They were chosen "arbitrarily" 😄. I didn't want the cluster to be overloaded as we plan to add sqlancer (which will take most of the cluster's capacity, I'm assuming). That said, we can increased these numbers for now and lower them in the future as well. I'll play around with them until I reach a workload that will reach ~50% CPU utilization.

srosenberg replied:
Makes sense, thanks. During the cloud report, we seemed to have found a way to _linearly_ scale cpu utilization via `tpcc-nowait` by increasing/decreasing connections and warehouses per vCPUs. Unfortunately, I wasn't able to reproduce it with 24.1.0. In any case, there are other ways to scale cpu utilization, albeit artificial, e.g., `stress-ng`.

srosenberg commented on pkg/cmd/roachtest/tests/mixed_version_multi_region.go:
```diff
@@ -0,0 +1,155 @@
+				mixedversion.AlwaysUseLatestPredecessors,
+			)
+
+			backgroundTPCCOpts := tpccOptions{
```
It might be worth mentioning explicitly that the background workload has no expiration (i.e., `Duration` is elided). It may not be immediately obvious given all the other config. options.

renatolabs replied:
Good point, added a comment about this.

srosenberg reviewed (approved):
Nice work!

renatolabs commented:
FWIW, I left this test running over night and it failed with an ambiguous error while in mixed version 23.2 and master. While workload should arguably handle that, the cluster had plenty of unused capacity at that point and I could see many suspicious error messages in the logs that would be worth looking into anyway.

Merging the test as-is so that any failures that do happen in the future get some eyes.

renatolabs commented:
bors retry

renatolabs commented:
bors retry

rickystewart commented:
bors r-

Build failures seem related to this PR

Please check the build logs before asking bors to retry to at least determine that it seems to be a flake. Bors has not made any progress today due to this.

renatolabs commented:
I thought I had fixed it in my latest force-push. Am I missing something?

renatolabs commented:
bors retry

renatolabs commented:
That didn't seem to do it, bors didn't move to Pending.

bors r=srosenberg,herkolategan
